ESQL: Union Types Support

craigtaverner opened the pull request:
If the query sources multiple indexes, and the same field exists in multiple indexes with different types, this would normally fail the query. However, if the query includes a conversion function to resolve the field to a single type before it is used in other functions or aggregations, then this should work.

The following query works in this third prototype:

```
FROM sample_data* METADATA _index
| EVAL client_ip = TO_IP(client_ip)
| KEEP _index, @timestamp, client_ip, event_duration, message
| SORT _index ASC, @timestamp DESC
```

The client_ip field is an `IP` in the `sample_data` index, but a `keyword` in the `sample_data_str` index.

The [first prototype](https://github.com/elastic/elasticsearch/pull/106885) did stuff to the drivers to create an index specific DriverContext to use during field evaluator construction so that the conversion function would be index/type aware. However, that abuses the idea of multi-threaded drivers. So the [second prototype](https://github.com/elastic/elasticsearch/pull/107255) took a new approach to instead re-plan the logical plan to extract the converter from the EVAL expressions, setting them as resolved (claiming the input type is already the converted type), and stored the converter in the EsRelation for later use in Physical planning. This third prototype takes this further by replacing the conversion function with a new FieldAttribute. ANd both old and new FieldAttributes exist in parallel, so that the logic around handling unsupported fields is not changed.

Fixes #100603 

Tasks to do:

* [x] Develop prototype that plans type conversion from the eval to the field extraction
* [x] Cleanup classes and types
  * [x] Move MultiTypeField to esql/types
  * [x] Add serialization of MultiTypeField to PlanNamedTypes for multi-node cluster support
* [x] Generate correct error messages when plan cannot be resolved (mimic previous behaviour)
* [x] Change solution to always create a new field (same name, different id) to not clash with existing field semantics
* [ ] Support rowStrideReader (using a conversion function directly in the ValueSourceReaderOperator)
  * [x] in `LoadFromMany`
  * [x] in `loadFromSingleLeaf`
* [x] Support conversion expression in any part of the plan (not just EVAL)
* [ ] More tests cases
   * [x] get union_types.csv-spec to work in CsvTests (or disable like ENRICH)
   * [x] Test with conversion function in WHERE clause
   * [x] test more complex cases (multiple evals, convert types back and forth, with and without stats)
   * [x] Unit tests for ValueSourceReaderOperator enhancements
   * [ ] tests covering row-stride-reader (synthetic source?)
* [ ] Bugs
  * [x] Fields with type `Date` are not working 
  * [x] Using invalid convert functions leads to incorrect error message (get runtime error instead of validation error)
  * [x] With no `KEEP` we get multiple columns with the same name
* [ ] https://github.com/elastic/elasticsearch/pull/110183

craigtaverner committed:
Union Types Support

The second prototype replaced MultiTypeField.Unresolved with MultiTypeField, but this clashed with existing behaviour around mapping unused MultiTypeFields to `unsupported` and `null`, so this new attempt simply adds new fields, resulting in more than one field with the same name.
We still need to store this new field in EsRelation, so that physical planner can insert it into FieldExtractExec, so this is quite similar to the second protototype.

The following query works in this third prototype:

```
multiIndexIpString
FROM sample_data* METADATA _index
| EVAL client_ip = TO_IP(client_ip)
| KEEP _index, @timestamp, client_ip, event_duration, message
| SORT _index ASC, @timestamp DESC
```

As with the previous prototyep, we no longer need an aggregation to force the conversion function onto the data node, as the 'real' conversion is now done at field extraction time using the converter function previously saved in the EsRelation and replanned into the EsQueryExec.

Support row-stride-reader for LoadFromMany

Add missing ESQL version after rebase on main

Fixed missing block release

Simplify UnresolvedUnionTypes

Support other commands, notably WHERE

Update docs/changelog/107545.yaml

Fix changelog

Removed unused code

Slight code reduction in analyser of union types

Removed unused interface method

Fix bug in copying blocks (array overrun)

Convert MultiTypeEsField.UnresolvedField back to InvalidMappedField

This is to ensure older behaviour still works.

Simplify InvalidMappedField support

Rather than complex code to recreate InvalidMappedField from MultiTypeEsField.UnresolvedField, we rely on the fact that this is the parent class anyway, so we can resolve this during plan serialization/deserialization anyway. Much simpler

Simplify InvalidMappedField support further

Combining InvalidMappedField and MultiTypeEsField.UnresolvedField into one class simplifies plan serialization even further.

InvalidMappedField is used slightly differently in QL

We need to separate the aggregatable used in the original really-invalid mapped field from the aggregatable used if the field can indeed be used as a union-type in ES|QL.

Updated version limitation after 8.14 branch

Try debug CI failures in multi-node clusters

Support type conversion in rowstride reader on single leaf

Disable union_types from CsvTests

Keep track of per-shard converters for LoadFromMany

Simplify block loader convert function

Code cleanup

Added unit test for ValuesSourceReaderOperator including field type conversions at block loading

Added test for @timestamp and fixed related bug

It turns out that most, but not all, DataType values have the same esType as typeName, and @timestamp is one that does not, using `date` for esType and `datetime` for typename. Our EsqlIndexResolver was recording multi-type fields with `esType`, while later the actual type conversion was using an evaluator that relied on DataTypes.typeFromName(typeName).
So we fixed the EsqlIndexResolver to rather use typeName.

Added more tests, with three indices combined and two type conversions

Disable lucene-pushdown on union-type fields

Since the union-type rewriter replaced conversion functions with new FieldAttributes, these were passing the check for being possible to push-down, which was incorrect. Now we prevent that.

Set union-type aggregatable flag to false always

This simplifies the push-down check.

Fixed tests after rebase on main

Add unit tests for union-types (same field, different type)

Remove generic warnings

Test code cleanup and clarifying comments

Remove -IT_tests_only in favor of CsvTests assumeFalse

Improved comment

Code review updates

Code review updates

Remove changes to ql/EsRelation

And it turned out the latest version of union type no longer needed these changes anyway, and was using the new EsRelation in the ESQL module without these changes.

Port InvalidMappedField to ESQL

Note, this extends the QL version of InvalidMappedField, so is not a complete port. This is necessary because of the intertwining of QL IndexResolver and EsqlIndexResolver. Once those classes are disentangled, we can completely break InvalidMappedField from QL and make it a forbidden type.

Fix capabilities line after rebase on main

Revert QL FieldAttribute and extend with ESQL FieldAttribute

So as to remove any edits to QL code, we extend FieldAttribute in the ESQL code with the changes required, since is simply to include the `field` in the hascode and equals methods.

Revert "Revert QL FieldAttribute and extend with ESQL FieldAttribute"

This reverts commit 168c6c75436e26b83e083cd3de8e18062e116bc9.

Switch UNION_TYPES from EsqlFeatures to EsqlCapabilities

Make hashcode and equals aligned

And removed unused method from earlier union-types work where we kept the NodeId during re-writing (which we no longer do).

Replace required_feature with required_capability after rebase

Switch union_types capability back to feature, because capabilities do not work in mixed clusters

Revert "Switch union_types capability back to feature, because capabilities do not work in mixed clusters"

This reverts commit 56d58bedf756dbad703c07bf4cdb991d4341c1ae.

Added test for multiple columns from same fields

Both IP and Date are tested

Fix bug with incorrectly resolving invalid types

And added more tests

Fixed bug with multiple fields of same name

This fix simply removes the original field already at the EsRelation level, which covers all test cases but has the side effect of having the final field no-longer be unsupported/null when the alias does not overwrite the field with the same name.
This is not exactly the correct semantic intent.
The original field name should be unsupported/null unless the user explicitly overwrote the name with `field=TO_TYPE(field)`, which effectively deletes the old field anyway.

Fixed bug with multiple conversions of the same field

This also fixes the issue with the previous fix that incorrectly reported the converted type for the original field.

More tests with multiple fields and KEEP/DROP combinations

Replace skip with capabilities in YML tests

Fixed missing ql->esql import change afer merging main

Merged two InvalidMappedField classes

After the QL code was ported to esql.core, we can now make the edits directly in InvalidMappedField instead of having one extend the other.

Move FieldAttribute edits from QL to ESQL

ESQL: Prepare analyzer for LOOKUP (#109045)

This extracts two fairly uncontroversial changes that were in the main
LOOKUP PR into a smaller change that's easier to review.

ESQL: Move serialization for EsField (#109222)

This moves the serialization logic for `EsField` into the `EsField`
subclasses to better align with the way rest of Elasticsearch works. It
also switches them from ESQL's home grown `writeNamed` thing to
`NamedWriteable`. These are wire compatible with one another.

ESQL: Move serialization of `Attribute` (#109267)

This moves the serialization of `Attribute` classes used in ESQL into
the classes themselves to better line up with the rest of Elasticsearch.

ES|QL: add MV_APPEND function (#107001)

Adding `MV_APPEND(value1, value2)` function, that appends two values
creating a single multi-value. If one or both the inputs are
multi-values, the result is the concatenation of all the values, eg.

```
MV_APPEND([a, b], [c, d]) -> [a, b, c, d]
```

~I think for this specific case it makes sense to consider `null` values
as empty arrays, so that~ ~MV_APPEND(value, null) -> value~ ~It is
pretty uncommon for ESQL (all the other functions, apart from
`COALESCE`, short-circuit to `null` when one of the values is null), so
let's discuss this behavior.~

[EDIT] considering the feedback from Andrei, I changed this logic and
made it consistent with the other functions: now if one of the
parameters is null, the function returns null

[ES|QL] Convert string to datetime when the other size of an arithmetic operator is date_period or time_duration (#108455)

* convert string to datetime when the other side of binary operator is temporal amount

ESQL: Move `NamedExpression` serialization (#109380)

This moves the serialization for the remaining `NamedExpression`
subclass into the class itself, and switches all direct serialization of
`NamedExpression`s to `readNamedWriteable` and friends. All other
`NamedExpression` subclasses extend from `Attribute` who's serialization
was moved ealier. They are already registered under the "category class"
for `Attribute`. This also registers them as `NamedExpression`s.

ESQL: Implement LOOKUP, an "inline" enrich (#107987)

This adds support for `LOOKUP`, a command that implements a sort of
inline `ENRICH`, using data that is passed in the request:

```
$ curl -uelastic:password -HContent-Type:application/json -XPOST \
    'localhost:9200/_query?error_trace&pretty&format=txt' \
-d'{
    "query": "ROW a=1::LONG | LOOKUP t ON a",
    "tables": {
        "t": {
            "a:long":     [    1,     4,     2],
            "v1:integer": [   10,    11,    12],
            "v2:keyword": ["cat", "dog", "wow"]
        }
    },
    "version": "2024.04.01"
}'
      v1       |      v2       |       a
---------------+---------------+---------------
10             |cat            |1
```

This required these PRs: * #107624 * #107634 * #107701 * #107762 *

Closes #107306

parent 32ac5ba755dd5c24364a210f1097ae093fdcbd75
author Craig Taverner <craig@amanzi.com> 1717779549 +0200
committer Craig Taverner <craig@amanzi.com> 1718115775 +0200

Fixed compile error after merging in main

Fixed strange merge issues from main

Remove version from ES|QL test queries after merging main

Fixed union-types on nested fields

Switch to Luigi's solution, and expand nested tests

Cleanup after rebase

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/MultiTypeEsField.java:
```diff
@@ -0,0 +1,125 @@
+     * If type resolution is not possible, due to the plan not containing explicit type conversion functions, then this class will
+     * be used to communicate the type resolution failure, since it's parent class is InvalidMappedField.
+     */
+    public static class UnresolvedField extends InvalidMappedField {
```
Why not use InvalidMappedField directly since that is already picked up by the analyzer/verifier.

craigtaverner replied:
This conversation has come full circle now, due to the recent reviews asking to not edit QL directly. After the above comment, I switched to editing QL directly. There were two classes where we edited QL `FieldAttribute` where the equals and hashcode were updated to include the underlying field, so query plan re-writing would work (could be seen as an oversight in QL code, so perhaps even a bug-fix), and the above mentioned InvalidMappedField. An attempt to completely extricate these two from QL failed (dependencies increased scope dramatically, turning a moderate PR into a monster). A second attempt at a compromise by making ESQL versions that maintained the simple name, and extended the QL versions worked for both classes, but was quite messy for FieldAttribute. After discussing with the team, we decided to keep the ESQL port of InvalidMappedField, but revert to the direct edit of the FieldAttribute. The thinking here is the up-coming split of QL and ESQL would be slightly facilitated by not having any edits in InvalidMappedField, while the edit to FieldAttribute we would probably want to keep anyway.

costin reviewed (commented):
Left a round of comments.

costin reviewed (commented):
Left a round of comments.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:
```diff
@@ -310,6 +311,13 @@ public static List<PlanNameRegistry.Entry> namedTypeEntries() {
             of(EsField.class, DateEsField.class, PlanNamedTypes::writeDateEsField, PlanNamedTypes::readDateEsField),
             of(EsField.class, InvalidMappedField.class, PlanNamedTypes::writeInvalidMappedField, PlanNamedTypes::readInvalidMappedField),
+            of(EsField.class, MultiTypeEsField.class, PlanNamedTypes::writeMultiTypeEsField, PlanNamedTypes::readMultiTypeEsField),
+            of(
+                EsField.class,
+                MultiTypeEsField.UnresolvedField.class, // This extends InvalidMappedField with information needed only during Analysis
+                PlanNamedTypes::writeInvalidMappedField,
+                PlanNamedTypes::readInvalidMappedField
+            ),
```
Agreed. It did actually fix a bunch of things, because `UnresolveUnionTypes` was becoming quite complex in order to get regressions to pass, and putting this here made it much simpler. However, if I instead combine the classes `InvalidMappedField` and `MultiTypeEsField.UnresolvedField` into one class, that same simplification occurs, without this hack. I noticed a lot of failures in CI that I could not reproduce locally, but this particular hack feels like a likely source of the failures.

craigtaverner replied:
Also, in case it was not clear, the existence of this in the plan was only when there is no AbstractConvertFunction, so the remaining InvalidMappedField needs to be serialised in the plan, to support existing behaviour around unsupported types.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -888,4 +893,127 @@ public static Expression castStringLiteral(Expression from, DataType target) {
+     * then there could still be some FieldAttributes that contain unresolved MultiTypeEsFields.
+     * These need to be converted back to actual UnresolvedAttribute in order for validation to generate appropriate failures.
+     */
+    private static class UnresolveUnionTypes extends AnalyzerRules.AnalyzerRule<LogicalPlan> {
```
This rule does something that used to be done by the ResolveRefs code. We have a catch-22 situation where ResolveUnionTypes depends on ResolveRefs to run first, but also relies on InvalidMappedField (or currently MultiTypeEsField.UnresolvedField) to not be converted into an UnresolvedAttribute (which ResolveRefs also does). So that conversion needed to be removed from ResolveRefs and moved after ResolveUnionTypes. This is that new location. This code also did something much more complex before, also editing the contents of EsRelation and the contained EsIndex, but that was getting very messy, and only needed to handle regression tests around unsupported types. I had three choices:

* Continue to make UnresolveUnionTypes more complex (I got all but on regression test passing, so just needed to add nested resolution of Object fields)
* Find a way to get MultiTypeEsField.UnresolvedField to pass through Plan serialization (this is the route I took, but it looks hacky, and I see your comment on that already). This allowed me to reduce this class a lot (you see the simplified version here).
* Merge MultiTypeEsField.UnresolvedField with InvalidMappedField so we avoid the plan serialization hack. This is my new preferred approach, and I see you suggest this too in another comment.

But in no cases does this class completely disappear. Somewhere we need to recognise InvalidMappedField as an unresolved type. That used to happen in ResolveRefs, and now happens here. You are suggesting an alternative location? Verifying EsqlProject? I can investigate that on Monday. I've looked through the Verifier a few times, so I can imagine possibilities there.

costin replied:
Thanks - it's a minor point but you could incorporate the rule as part of the UnionType rules in a second pass.
That is convert all InvalidMappedFields that were are not wrapped in a conversion function to the unresolved attribute.
Again minor.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner commented:
After some recent refinements, and the addition of the unit tests in `ValuesSourceReaderTypeConversionTests`, I think it is time for some reviews!

luigidellaquila reviewed (commented):
Hi Craig, I did a couple of tests, but I think I need some guidance to understand which types/functions this PR covers, because I couldn't manage to make it work...

I did the following:
- created two indexes, `test1` and `test2` with only one `@timestamp` field of type `long` and `date` respectively. 
- tried a few queries with empty indexes:
  - `from test* | eval x = to_long(@timestamp) | keep x`: all good
  - `from test* | eval x = to_string(@timestamp) | keep x`: all good
  - `from test* | eval x = to_date(@timestamp) | keep x`: `verification_exception`, `Found 1 problem\nline 1:31: Cannot use field [@timestamp] due to ambiguities being mapped as [2] incompatible types: [date] in [test2], [long] in [test1]`

Then I added one record per index

```
{"index": {"_index":"test1"}}
{"@timestamp":10000000}
{"index": {"_index":"test2"}}
{"@timestamp":"2022-05-06T12:01:00.000Z"}
```

and tried the same queries again:
- `from test* | eval x = to_long(@timestamp) | keep x`: `null_pointer_exception`, `Cannot invoke \"Object.hashCode()\" because \"pk\" is null`
- `from test* | eval x = to_string(@timestamp) | keep x`: same as above
- `from test* | eval x = to_date(@timestamp) | keep x`: `verification_exception`. `Found 1 problem\nline 1:31: Cannot use field [@timestamp] due to ambiguities being mapped as [2] incompatible types: [date] in [test2], [long] in [test1]`
- `from test* | eval x = to_ip(@timestamp) | keep x`: `esql_illegal_argument_exception`, `illegal data type [long]` (suppressed `null_pointer_exception` as above)

I also tried `IP` vs `long` and I got some results with `to_string()` but not with other functions.

So a couple of questions:
- does this cover all the types? I didn't find any type-specific code in the PR, so I assumed it was the case.
- is it supposed to handle partially invalid conversions? Eg. if I have an IP and a date and I'm only interested in the date, can I do it?

[edit] it should be `to_datetime()` (`to_date()` does not exist); probably the validation order makes it trip on the incompatible types before it realizes that the function name is wrong, it makes the message a bit confusing, but probably it's a minor problem. 
The above queries, with `to_datetime()` return the same error as `to_long()`

nik9000 commented:
I feel bad for the github user `@timestamp`. We ping them so much.

nik9000 commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/FieldAttribute.java:
```diff
@@ -142,7 +142,9 @@ public int hashCode() {
-        return super.equals(obj) && Objects.equals(path, ((FieldAttribute) obj).path);
+        return super.equals(obj)
+            && Objects.equals(path, ((FieldAttribute) obj).path)
+            && Objects.equals(field, ((FieldAttribute) obj).field);
```
Modifying this one scares me a bit.

alex-spies replied:
Instead of modifying this, we should create a copy in the esql project.

This may involve some yak shaving, but I'm happy to help with that!

nik9000 reviewed (changes requested):
I left a few comments. I suggested reworking how row-by-row loading is modeled with the `BlockBuilder`s inside of the value loaders themselves. This *feels* like it's much cleaner to think about. I think we can make it compatible with a shared loader if we need to. But for now the copying seems fine because this is never the hottest path.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/InvalidMappedField.java:
Clone this to ESQL and add the modifications there.

craigtaverner replied:
OK. I'll discuss this with @alex-spies and do it together with the FieldAttribute changes, since both are really part of the 'split from QL' project.

craigtaverner replied:
Done

costin reviewed (commented):
I have minor comments around the logistics and code dependency - such as moving the QL classes like FieldAttribute, EsRelation and InvalidMapperField to ESQL.
An alternative would be to have some workaround code in ESQL that adds/maintains that information outside of QL (even though it's ugly) and then get rid of that during the QL migration to not postpone this PR any longer.

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -335,16 +336,20 @@ void run() throws IOException {
-                builders[f] = fields[f].info.type.newBlockBuilder(docs.getPositionCount(), blockFactory);
+                fieldTypeBuilders[f] = fields[f].info.type.newBlockBuilder(docs.getPositionCount(), blockFactory);
+                builders[f] = new Block.Builder[shardContexts.size()];
+                converters[f] = new BlockLoader[shardContexts.size()];
```
I remember originally trying the combine these, but seem to remember it causing hassles with the ColumnAtATimeReader. This was before I really got the row-stride reader working, so perhaps it is time for another attempt.

nik9000 replied:
I'd be ok delaying it to a followup too.

craigtaverner replied:
I realise a key difference between what you suggested and what I tried (and was thinking of trying again). You said to put the builder and converter inside the loader. Instead I had tried to put the converter inside the builder. This is because we will always have both a builder and loader (the builder is made by the loader once additional information is known using code like `field.loader.builder(loaderBlockFactory, docs.count())`), so merging the loader and builder seems unworkable. Right now the converter is the only thing at play. It is currently implemented by the loader, since it does not need any of the information in the builder. But this does lead to lines like this:

```
Block build() {
    return (Block) loader.convert(builder.build());
}
```

If I put the converter into the builder, we can hide it inside the `build()` method call, simplifying things slightly. That was what I was trying before, and could try again.

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types-IT_tests_only.csv-spec:
```diff
@@ -0,0 +1,582 @@
+singleIndexIp
```
OK. I've done this now, will push soon.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsPhysicalOperationProviders.java:
```diff
@@ -101,17 +113,42 @@ public final PhysicalOperation fieldExtractPhysicalOperation(FieldExtractExec fi
+            String indexName = shardContext.ctx.index().getName();
+            Expression conversion = unionTypes.getConversionExpressionForIndex(indexName);
+            return new TypeConvertingBlockLoader(blockLoader, (AbstractConvertFunction) conversion);
+        }
```
Yeah, I think I had an earlier prototype where the union-types contained multiple fields, so this probably inherited from that. I'm removing the check.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsPhysicalOperationProviders.java:
```diff
@@ -320,4 +357,96 @@ public FieldNamesFieldMapper.FieldNamesFieldType fieldNames() {
+
+        @Override
+        public final String toString() {
+            return "TypeConvertingBlockLoader[delegate=" + delegate + "]";
```
Done

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsPhysicalOperationProviders.java:
```diff
@@ -320,4 +357,96 @@ public FieldNamesFieldMapper.FieldNamesFieldType fieldNames() {
+                }
+
+                @Override
+                public void close() {}
```
Unfortunately the `block.incRef();` causes all integration tests to fail with `estimated_size_in_bytes: expected <0> but was <256>`

nik9000 replied:
Ah, that makes sense! Ignore it then. Could you add a `toString` to this one that's at least similar? I find I really like to have a nice `toString` on these things.

craigtaverner replied:
I could add a toString, but I note that the other one does not have a toString. Only the factory does, and we are not using a factory here. Our factory is a lambda, and perhaps you want me to convert that into an actual Factory class?

craigtaverner replied:
Hmmm... I tried this, but it made the code a lot more obtuse. I think the fact that we create the factory and immediately call the get method on it, within the very same statement, means it has no intrinsic value. The toString will never be called, and the extra boiler-plate obsfucates the code. I reverted back to a lambda.

This piece of code is designed to extract the Evaluators normally used within a complete execution pipeline in a complex plan, and inject it into the BlockLoader. All this driverContext/EvaluatorFactory/Evaluator stuff makes sense in the normal execution pipeline, but is just additional ritual in this case. Unavoidable ritual. Except in the unit tests, where I create some hard-coded evaluators and skip all this.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlFeatures.java:
```diff
@@ -104,6 +104,10 @@ public class EsqlFeatures implements FeatureSpecification {
+    /**
+     * Support multiple field mappings if appropriate conversion function is used (union types)
+     */
+    public static final NodeFeature UNION_TYPES = new NodeFeature("esql.union_types");
```
This sounds like a solution to a problem we suspect we might get in future, but do not (yet) suffer from, the issue of the creation of too many NodeFeatures. I think there are many solutions to that problem, like regular removal of older NodeFeature instances and references to them, since they should only really matter for BWC and rollover of recent releases. Or some other solution... let's not make it part of this PR (or any feature PR).

nik9000 replied:
I'm with @craigtaverner on this. I'd prefer not to adopt a new pattern here. I'm not really sure what the pattern should be either. Or if we need one.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlFeatures.java:
```diff
@@ -153,6 +157,7 @@ public Set<NodeFeature> getFeatures() {
             ST_CONTAINS_WITHIN,
             ST_DISJOINT,
             STRING_LITERAL_AUTO_CASTING,
+            UNION_TYPES,
```
That's pretty reasonable.

nik9000 replied:
Maybe just stick this one on top and leave a comment saying add new ones on top.

nik9000 replied:
This is just for those reading - it's an unordered Set - but we humans do read it in order.

craigtaverner replied:
So far it seems everyone, including me, has be appending to the end. So we're talking about reversing the order of the complete list? I can do that, although I don't fully understand the reason for this. The term 'wildly spread' does not clarify things for me.

alex-spies commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/logical/EsRelation.java:
```diff
@@ -96,12 +96,14 @@ public List<Attribute> output() {
     @Override
     public int hashCode() {
-        return Objects.hash(index, esSourceOptions, frozen);
+        return Objects.hash(index, attrs, esSourceOptions, frozen);
```
`EsRelation` was already ported to `esql` by Nhat - I think this fix should be applied to `esql`'s copy of that, not here.

craigtaverner replied:
Curiously, in the QL version, it used attrs in the hashCode, but not in the equals method. My change adds it to equals. Nhat's version removed it from hashCode. I'm curious if his change fixed something else, that I will now break.

craigtaverner replied:
Turns out the changes I made were no longer necessary, and only needed for an earlier version of union-types. I kept just one clarifying comment.

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/InvalidMappedField.java:
```diff
@@ -0,0 +1,74 @@
+ * TODO: Once QL and ESQL code bases are split, we should incorporate the error message from the QL version into this class.
+ * Breaking the inheritance requires that we port IndexResolver to ESQL, so it cannot be done yet.
+ */
+public class InvalidMappedField extends org.elasticsearch.xpack.ql.type.InvalidMappedField {
```
This scares me a bit TBH...
Eg. we still use QL IndexResolver that returns a QL `InvalidMappedField`. 
Could it for instance fail de/serialization, since `PlanNamedTypes` only knows the ES|QL version of `InvalidMappedField`?

craigtaverner replied:
InvalidMappedField does not need to survive serialisation, and the changes are purely additive. So in my opinion, either approach works as well:

* Edit InvalidMappedField directly and add a comment about the additions being purely for ESQL use cases
* Extend InvalidMappedField with the ESQL version, so QL code is left untouched, and this is a step towards the QL/ESQL split.

In both cases, the common behaviour is identically serializable, and the information lost when serializating the ESQL version to QL is of no consequence. I can add a comment about that.

craigtaverner replied:
We took the latter approach, since it is a step towards splitting from QL. We did this for InvalidMappedField, but not for FieldAttribute, since that was much, much messier to port, and the changes were seen as an improvement anyway.

luigidellaquila reviewed (commented):
Left a comment about duplicating QL structures, I'm wondering if there could be collisions, eg. on serialization

nik9000 reviewed (approved):
I'm happy with it. I might try to refactor the conversion a bit but I'll give that a go in a follow up. When the QL folks are happy I'm happy.

luigidellaquila commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -0,0 +1,594 @@
+required_capability: union_types
+
+FROM sample_data*
+| WHERE TO_LONG(@timestamp) < 1698068014937 AND TO_STRING(client_ip) LIKE "172.21.2.16?"
```
If you have a chance, please add a few tests where you convert the same field multiple times to different types, eg.

```
| EVAL x = to_string(@timestamp), y = to_datetime(@timestamp), z= to_long(@timestamp)
```

I tested it and it works fine btw, but it's worth having the coverage ;-)

luigidellaquila replied:
And I think conversions to incompatible types (eg. `to_ip(date)`) have some validation problems. Nothing critical, but it's worth checking.

Probably in these cases we can just return null (we should probably, at least when one of the types is compatible with the conversion), but we could also fail the validation phase with a meaningful error message, up to you.

luigidellaquila replied:
One more thing that just came to my mind:

with multiple indexes with different `@timestamp` fields:

```
from test* | drop @timestamp
```

works just fine

```
from test* | eval x = to_datetime(@timestamp), y = to_string(@timestamp) | drop x, y, @timestamp
```

returns a verification exception:

```
{
    "error": {
        "root_cause": [
            {
                "type": "verification_exception",
                "reason": "Found 2 problems\nline 1:1: Cannot use field [@timestamp] due to ambiguities being mapped as [3] incompatible types: [datetime] in [test2], [keyword] in [test3, test4], [long] in [test1]\nline 1:87: Reference [@timestamp] is ambiguous (to disambiguate use quotes or qualifiers); matches any of [line 1:1 [@timestamp], line 1:35 [@timestamp], line 1:62 [@timestamp]]"
            }
        ],
        "type": "verification_exception",
        "reason": "Found 2 problems\nline 1:1: Cannot use field [@timestamp] due to ambiguities being mapped as [3] incompatible types: [datetime] in [test2], [keyword] in [test3, test4], [long] in [test1]\nline 1:87: Reference [@timestamp] is ambiguous (to disambiguate use quotes or qualifiers); matches any of [line 1:1 [@timestamp], line 1:35 [@timestamp], line 1:62 [@timestamp]]"
    },
    "status": 400
}
```

I think it's due to the fact that now you have multiple `@timestamp` fields.

Also, 

```
from test* | eval x = to_datetime(@timestamp), y = to_string(@timestamp)
```

actually returns multiple `@timestamp` fields (with different types), including the original `unsupported`. I don't think it's what we want... we should at least implicitly them after the renaming, otherwise basic operations like a DROP will become difficult.

luigidellaquila replied:
> so we generate the correct error at planning (validation) time.

I wonder if it makes sense to just short-circuit the wrong conversions, returning null. It would allow to manage field name collisions where a single conversion cannot be applied (eg. when the dataset contains userful data + garbage)

craigtaverner replied:
Fixing the wrong conversion function was easy enough. The problem was just that when resolving whether a multi-type field mapping was satisfied by an existing convert function, I never checked if the convert function actually supported those types. I just needed to add that check. Then the existing code paths for invalid mapped fields would kick in as normal.

craigtaverner replied:
But the third issue is more subtle. The idea was to generate a new field with the same name as the old one, and the old one would satisfy use cases where we expect to see `unsupported` and `null`, while the new one would satisfy the use cases where we expect the type to by converted. This seems to have worked great for a wide number of cases, but fails on the `DROP field` command, which I had not tested. It seems like we need to update `DROP` to handle this case. I don't want to remove the new field, because it still covers the case where we do not DROP, and we expect to see the extra field.

luigidellaquila replied:
I think we'll have to hide the newly generated fields anyway, even if we don't have a DROP in the query.

Consider the following:

luigidellaquila replied:
I think the new fields have to be dropped anyway, even without a DROP

Consider the following:

```
PUT index1
{
    "mappings": {
        "properties": {
            "@timestamp": {
                "type": "long"
            },
            "name": {
                "type": "keyword"
            }
        }
    }
}

PUT index2
{
    "mappings": {
        "properties": {
            "@timestamp": {
                "type": "date"
            },
            "name": {
                "type": "keyword"
            }
        }
    }
}
```

```
from test* | eval x = to_datetime(@timestamp)
```

```
{
    "columns": [
        {
            "name": "@timestamp",
            "type": "unsupported"
        },
        {
            "name": "name",
            "type": "keyword"
        },
        {
            "name": "@timestamp",
            "type": "date"
        },
        {
            "name": "x",
            "type": "date"
        }
    ],
    "values": []
}
```

We have two `@timestamp` fields now, that unexpected

craigtaverner replied:
OK, I've written a few more tests, and indeed if we don't explicitly type `KEEP` we do end up with multiple columns with the same name, as you suspected. So likely the fix to that will also fix the `DROP` error. All my existing tests used `KEEP` to ensure predictable column ordering, but it turns out they had the side effect of hiding this issue.

luigidellaquila replied:
> I don't want to remove the new field, because it still covers the case where we do not DROP, and we expect to see the extra field.

As a user, I expect to only see the value of `x` in this case; the extra `@timestamp` field is an implementation detail that should not appear in the result IMHO

craigtaverner replied:
It is fixed in the latest commits, including tests that assert the new fields are not in the final output.

luigidellaquila reviewed (commented):
I left a couple of comments on tests, probably it needs just a little fix on the validation side (or to make a bit more resilient to invalid conversions) [edit] and a bit more logic to avoid leaking the newly added fields. [/edit]

Apart from that, technically it works, so it LGTM, but please wait for a final approval from @costin on the general approach.

As we discussed, there are two aspects that don't completely convince me, but as long as we are aware of them and we all agree, I'm good with them as well:
1. having `MultiTypeEsField` contain `Expression`s does not feel natural (thought it's functional to the solution we adopted)
2. doing the conversion at extraction time could be expensive, and since we can no longer decide where to push the execution of the conversion, we cannot speculate on it in the optimizer

I would have preferred a solution where you:
- replace the original multi-type field with multiple (normal) fields, one per type
- extract the new fields normally, just multiplexing on the right type (and using NullBlocks where the types don't match, that should be pretty efficient in terms of both memory and performance)
- replace the conversion function with an optimized version of `COALESCE(newField1, to_xy(newField2), ...)`, that can be further optimized (and sometimes maybe completely removed) by the optimizers

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
I don't see the basic test I mentioned some time ago that shows the union type and its various facets at the same time:
```
FROM sample_data
EVAL as_ip = TO_IP(client_ip), as_string = TO_STRING(client_ip)
```
Same for `@timestamp`.

craigtaverner replied:
These tests were written in the file 140_union_types.yml where it was easier to create more than just two index mappings, and also where I could write negative tests (tests asserting correct errors were thrown). The tests covered combinations of date, IP and long types, and used functions like `TO_IP` and `TO_LONG`, separately and together, but did not cover `@timestamp` conversion. Instead that was covered in `union_types.csv-spec` using the mapping `sample_data_ts.json`, where I combined `TO_DATETIME` with `TO_IP`. I also have tests with `TO_STRING`, and nested functions like `TO_STRING(TO_IP(...))`.

craigtaverner replied:
I just noticed that one difference between the tests I wrote and what you are asking for is the creation of multiple columns from the same original field. I'm writing a test for that now.

craigtaverner replied:
OK. I've pushed a test that creates multiple columns from the same field.

costin reviewed (commented):
I might have missed but I don't see the basic test (see my previous comment):
```
FROM sample_data
EVAL as_ip = TO_IP(client_ip), as_string = TO_STRING(client_ip)
```

or 
```
FROM sample_data
EVAL as_ts = TO_DATETIME(@timestamp), as_string = TO_STRING(@timestamp)
```

costin commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValueSourceReaderTypeConversionTests.java:
```diff
@@ -0,0 +1,2021 @@
+                newIndexWriterConfig().setMergePolicy(NoMergePolicy.INSTANCE).setMaxBufferedDocs(IndexWriterConfig.DISABLE_AUTO_FLUSH)
+            )
+        ) {
+            for (int d = 0; d < size; d++) {
```
This code is highly specific to this test - it would be useful to have a description of what the docs look like since there's no templating or external file to look at.

costin commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValueSourceReaderTypeConversionTests.java:
Maintaining a 2K file is going to be a challenge - it's over half of the PR.
In time we should try and externalize the field extractor creation - but that's for another time.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -0,0 +1,644 @@
+required_capability: metadata_fields
+
+FROM sample_data, sample_data_str METADATA _index
+| EVAL host_ip = TO_STRING(TO_IP(client_ip))
```
👍

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -0,0 +1,644 @@
+required_capability: metadata_fields
+
+FROM sample_data, sample_data_str METADATA _index
+| WHERE STARTS_WITH(TO_STRING(client_ip), "172.21.2")
```
👍

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -0,0 +1,644 @@
+
+FROM sample_data* METADATA _index
+| EVAL @timestamp = TO_DATETIME(@timestamp), client_ip = TO_IP(client_ip)
+| KEEP _index, @timestamp, client_ip, event_duration, message
```
👍

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -928,4 +925,154 @@ public static Expression castStringLiteral(Expression from, DataType target) {
+
+            // Otherwise drop the converted attributes after the alias function, as they are only needed for this function, and
+            // the original version of the attribute should still be seen as unconverted.
+            plan = dropConvertedAttributes(plan, unionFieldAttributes);
```
This should not be needed - let's talk more on why this happens.

craigtaverner replied:
This was the fix to the one of the issues that Luigi reported. Without this, there were multiple fields with the same name in the output, the original field with unsupported type (and null values) and the converted field (or many converted fields if we have the same field converted many times), with their converted values. But in reality all the converted fields are immediate aliased to whatever was written in the query, so the internal converted fields need to be removed.

astefan replied:
If I understand this well, you remove these attributes assuming they are only used in the "alias function" by which you mean an `eval x = field::type`. I don't think `eval` is the only place where you can place such a conversion function.

`STATS c=count(*) BY client_ip::ip`

`FROM sample_data, sample_data_str | SORT client_ip::ip`

Both of these have issues. After you merge this PR, I will create follow-ups for these two.

costin reviewed (commented):
Thanks for the updates - 👍 that the majority of this PR is made up of tests.

astefan reviewed (commented):
Started looking at the PR and the first thing that jumped at me (and I stopped reviewing at that point to both have a hopefully useful review and also quick one) was the new FieldAttributes that get created... more specifically in the Analyzer `FieldAttribute resolvedAttr = new FieldAttribute(source, null, field.getName(), field, null, Nullability.TRUE, id, false);` (the `null` parent)

Hierarchy of fields in the QL ecosystem is tricky, but that's not the point. I think you shouldn't ignore that "parent" constructor argument and think of this as a whole new use case where the fields that can be converted are sub-fields of other fields.

An example below:

```
test1

{
  "mappings": {
    "properties": {
        "obj": {
            "properties": {
                "keyword": {
                    "type": "keyword"
                },
                "integer": {
                    "type": "integer"
                }
            }
        },
        "keyword": {
            "type": "keyword"
        },
        "integer": {
            "type": "integer"
        }
    }
  }
}
```

```
test2

{
  "mappings": {
    "properties": {
        "obj": {
            "properties": {
                "keyword": {
                    "type": "boolean"
                },
                "integer": {
                    "type": "version"
                }
            }
        },
        "keyword": {
            "type": "boolean"
        },
        "integer": {
            "type": "version"
        }
    }
  }
}
```

Test data

```
{ "index": {"_id": 1, "_index":"test1"} }
{ "obj.keyword": "true", "obj.integer": 100, "keyword": "true", "integer": 100 }
{ "index": {"_id": 2, "_index":"test1"} }
{ "obj.keyword": "US", "obj.integer": 20, "keyword": "US", "integer": 20 }
{ "index": {"_id": 12, "_index":"test2"} }
{ "obj.keyword": "true", "obj.integer": "50", "keyword": "true", "integer": "50" }
{ "index": {"_id": 22, "_index":"test2"} }
{ "obj.keyword": false, "obj.integer": "1.2.3", "keyword": false, "integer": "1.2.3" }
```


Query to test:
```
from test* metadata _id | eval v = to_version(obj.integer), s = to_string(obj.keyword) | keep v, s, _id
```

and even simpler

```
from test* metadata _id | eval s = to_string(obj.keyword) | keep s, _id
```

fails with 

```
NullPointerException: Cannot invoke "org.elasticsearch.xpack.esql.core.expression.Expression.dataType()" because "convertExpr" is null
        at org.elasticsearch.xpack.esql.type.MultiTypeEsField.resolveFrom(MultiTypeEsField.java:57)
        at org.elasticsearch.xpack.esql.analysis.Analyzer$ResolveUnionTypes.resolvedMultiTypeEsField(Analyzer.java:1038)
        at org.elasticsearch.xpack.esql.analysis.Analyzer$ResolveUnionTypes.resolveConvertFunction(Analyzer.java:1004)
        at org.elasticsearch.xpack.esql.analysis.Analyzer$ResolveUnionTypes.lambda$doRule$0(Analyzer.java:949)
```

while the following one works `from test* metadata _id | eval s = to_string(keyword) | keep s, _id`

luigidellaquila reviewed (commented):
Thanks @craigtaverner, the solution with the nested fields should cover all the cases.
I also tested it with a more subtle case, ie. a schema with a TEXT field and a non-indexed KEYWORD subfield, and it works fine. 
In case you want to add it to the suite, this is more or less what I added:

```
  - do:
      indices.create:
        index: weird
        body:
          mappings:
            properties:
              obj:
                type: text
                fields:
                  keyword:
                    type: "keyword"
                    index: false
                    doc_values: false
                    store: true
```
```

  - do:
      bulk:
        refresh: true
        index: weird
        body:
          - '{ "index": {"_id": 31} }'
          - '{ "obj": "1000"}'
```
```
---
"with nested non-indexed fields":
  - do:
      allowed_warnings_regex:
        - "No limit defined, adding default limit of \\[.*\\]"
      esql.query:
        body:
          query: 'FROM test*,weird METADATA _id | EVAL nv = TO_VERSION(TO_STRING(obj.keyword)) | KEEP _id, nv | SORT _id ASC'

  - match: { columns.0.name: "_id" }
  - match: { columns.0.type: "keyword" }
  - match: { columns.1.name: "nv" }
  - match: { columns.1.type: "version" }
  - length: { values: 5 }
  - match: { values.0.0: "11" }
  - match: { values.0.1: "true" }
  - match: { values.1.0: "12" }
  - match: { values.1.1: "US" }
  - match: { values.2.0: "21" }
  - match: { values.2.1: "true" }
  - match: { values.3.0: "22" }
  - match: { values.3.1: "false" }
  - match: { values.4.0: "31" }
  - match: { values.4.1: "1000" }
```

Re. the comment from @astefan regarding the parent set to `null`, it looks strange indeed. ESQL is not doing much with that hierarchy, so probably it's fine as it is, but I'd wait for a final confirmation from Andrei.

I only see a CI failure on `union_types.MultiIndexIpStringRenameStats` that could be a regression and is worth checking:

```

Caused by: java.lang.IllegalArgumentException: expected exactly [16] bytes but got [11]
--
  | at org.elasticsearch.compute.operator.topn.FixedLengthTopNEncoder.encodeBytesRef(FixedLengthTopNEncoder.java:23)
  | at org.elasticsearch.compute.operator.topn.KeyExtractorForBytesRef.nonNul(KeyExtractorForBytesRef.java:50)
  | at org.elasticsearch.compute.operator.topn.KeyExtractorForBytesRef$FromVector.writeKey(KeyExtractorForBytesRef.java:73)
  | at org.elasticsearch.compute.operator.topn.TopNOperator$RowFiller.writeKey(TopNOperator.java:183)
  | at org.elasticsearch.compute.operator.topn.TopNOperator$RowFiller.row(TopNOperator.java:176)
  | at org.elasticsearch.compute.operator.topn.TopNOperator.addInput(TopNOperator.java:360)
  | at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:255)
  | at org.elasticsearch.compute.operator.Driver.run(Driver.java:178)
  | at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:364)
  | ... 7 more
```

Apart from that, I think we could start considering to merge it.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -0,0 +1,644 @@
+2023-10-23T12:15:03.360Z  |  172.21.2.162  |  3450233              |  Connected to 10.1.0.3
+;
+
+singleIndexWhereIpLike
```
Please add a test where the converted field is not used / dropped:
```
from test
| eval my_client_ip = TO_IP(client_ip)
| keep @timestamp
```

And another one where the original name is overridden:
```
from test
| eval my_client_ip = TO_IP(client_ip)
| keep @timestamp
```

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1061,4 +1062,155 @@ public static Expression castStringLiteral(Expression from, DataType target) {
+            List<FieldAttribute> unionFieldAttributes = new ArrayList<>();
+            // See if the eval function has an unresolved MultiTypeEsField field
+            // Replace the entire convert function with a new FieldAttribute (containing type conversion knowledge)
+            plan = plan.transformExpressionsOnly(
```
Approaching in place is short and effective however it has a number of side-effects:
- creates new fields with the same name that clash the original source and need to be managed
- need to be pushed into the underlying source which creates additional input for previous commands. This issue might be hidden due to the fact that the field could not have been used before.
- the analyzer ends up setting up a payload for the fieldExtractor which is somewhat fragile and could be avoided

A better approach would be to extract these conversion fields into a synthetic eval right before its usage and then handle the pushdown of the converted valuesource during physical planning.

For example:
```
from index | where to_ip(client_ip) != "127.0.0.1"
```
becomes
```
from index | eval _ = to_ip(client_ip) | where _ != "127.0.0.1"
```

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -0,0 +1,644 @@
+;
+
+singleIndexIpStringStats
+FROM sample_data_str
+| EVAL client_ip = TO_IP(client_ip)
+| STATS count=count(*) BY client_ip
```
Please add a variant with the shorthand example:
```
from sample
| stats count = count(*) by client_ip = to_ip(client_ip)
...
```
And another one with the conversion inside the aggregation which later gets dropped:

```
from index
| stats m = max(to_string(client_ip)), c = count()
| keep c
```

costin reviewed (approved):
I think there's room for improvement however this PR took long enough and github doesn't like this commit history.
I'd wait for Andrei's comments, create a separate issue with feedback, merge this in and address the remaining issues in a separate PR.

Thanks for your patience Craig.

costin reviewed (approved):
I think there's room for improvement however this PR took long enough and github doesn't like this commit history.
I'd wait for Andrei's comments, create a separate issue with feedback, merge this in and address the remaining issues in a separate PR.

Thanks for your patience Craig.

costin reviewed (approved):
I think there's room for improvement however this PR took long enough and github doesn't like this commit history.
I'd wait for Andrei's comments, create a separate issue with feedback, merge this in and address the remaining issues in a separate PR.

Thanks for your patience Craig.

craigtaverner committed:
Added more tests from code review

Note that one test, `multiIndexIpStringStatsInline` is muted due to failing with the error:

    UnresolvedException: Invalid call to dataType on an unresolved object ?client_ip

astefan reviewed (approved):
Thank you for adding tests. I like them.
I've left few very minor comments and one that is more serious.
I think there is value in the PR and it should be merged, but it needs follow-ups.

LGTM.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into union_types_take3

luigidellaquila reviewed (approved):
LGTM

craigtaverner committed:
Make CsvTests consistent with integration tests for capabilities

The integration tests do not fail the tests if the capability does not even exist on cluster nodes, instead the tests are ignored. The same behaviour should happen with CsvTests for consistency.

craigtaverner committed:
Return assumeThat to assertThat, but change order

This way we don't have to add more features to the test framework in this PR, but we would probably want a mute feature (like a `skip` line).

craigtaverner committed:
Move serialization of MultiTypeEsField to NamedWritable approach

Since the sub-fields are AbstractConvertFunction expressions, and Expression is not yet fully supported as a category class for NamedWritable, we need a few slight tweaks to this, notably registering this explicitly in the EsqlPlugin, as well as calling PlanStreamInput.readExpression() instead of StreamInput.readNamedWritable(Expression.class). These can be removed later once Expression is fully supported as a category class.

craigtaverner committed:
Remove attempt to mute two failed tests

We used required_capability to mute the tests, but this caused issues with CsvTests which also uses this as a spelling mistake checker for typing the capability name wrong, so we tried to use muted-tests.yml, but that only mutes tests in specific run configurations (ie. we need to mute each and every IT class separately).

So now we just remove the tests entirely. We left a comment in the muted-tests.yml file for future reference about how to mute csv-spec tests.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into union_types_take3

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/MultiTypeEsField.java:
```diff
@@ -0,0 +1,121 @@
+            in.readString(),
+            DataType.readFrom(in),
+            in.readBoolean(),
+            in.readImmutableMap(StreamInput::readString, i -> ((PlanStreamInput) i).readExpression())
```
I tend to just do `readMap( i -> ((PlanStreamInput) i).readExpression())`

nik9000 replied:
That's a little more convenient and I just trust that we don't mutate the map.

Though I should probably do:
```
readImutableMap(i -> ((PlanStreamInput) i).readExpression())
```

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/MultiTypeEsField.java:
```diff
@@ -0,0 +1,121 @@
+        out.writeString(getName());
+        out.writeString(getDataType().typeName());
+        out.writeBoolean(isAggregatable());
+        out.writeMap(getIndexToConversionExpressions(), (o, v) -> out.writeNamedWriteable(v));
```
:+1:
This what we need. It lines up perfectly with `writeNamed` from the old stuff.

nik9000 replied:
Actually, I don't believe this is going to work *yet*. `Expression` isn't *yet* fully properly implemented as a `NamedWriteable`. You'd be better off doing
```
out.writeMap(getIndexToConversionExpressions(), (o, v) -> ((PlanStreamOutput) o).writeExpression(v))
```

I'm in the process of converting all remaining `Expression` subclasses to proper `NamedWriteable`s. they don't all properly implement it yet.

craigtaverner replied:
And yet it does work. At least my multi-node integration tests pass, and they rely on this serialisation succeeding (both for write and read). I assumed it was because the on-the-wire serialization was the same in both cases.

nik9000 replied:
It'll only work if we've implemented `NamedWriteable` for all the subclasses you want to serialize. I'm doing that with stuff like https://github.com/elastic/elasticsearch/pull/109892 but not all subclasses are done. I suppose if you only allow certain stuff in there it'll be fine. I dunno.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/MultiTypeEsField.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    public MultiTypeEsField(StreamInput in) throws IOException {
+        // TODO: Change the conversion expression serialization to i.readNamedWriteable(Expression.class) once Expression is fully supported
```
That TODO will get done before too long. I'll remove `readExpression` entirely.

nik9000 commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/161_union_types_nested.yml:
```diff
@@ -0,0 +1,203 @@
+setup:
```
The word "nested" means the `nested` type to me. Could you rename this file to something like "sub_fields" or something? I forgot what we call those sub field things, but `nested` and `object` mean something specific to me.

nik9000 reviewed (approved):
@craigtaverner told me that the concurrent serialization tests are taking a surprising amount of time on this PR so I'm going to check that.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into union_types_take3

craigtaverner committed:
Fix rather massive issue with performance of testConcurrentSerialization

Recreating the config on every test was very expensive.

craigtaverner committed:
Code review by Nik

elasticmachine committed:
Merge branch 'main' into union_types_take3

bpintea reviewed (commented):
Leaving some comments that could be considered for later.

bpintea reviewed (commented):
And a last batch of (post-merge) notes to potentially be later considered. 
Very nice.

craigtaverner commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/InvalidMappedField.java:
```diff
@@ -15,11 +15,15 @@
 /**
  * Representation of field mapped differently across indices.
  * Used during mapping discovery only.
+ * Note that the field <code>typesToIndices</code> is not serialized because that information is
```
There really should not be. This class, InvalidMappedField is used in the Analyzer to produce either UnsupportedField, or MultiTypeEsField (the non-union-types and the union-types), long before any serialization. The typesToIndices is used to produce the think that those two fields need, either and error message, or the union-types converters (which are serialized as part of MultiTypeEsField).

What would make sense in future, once the port to NamedWritable is complete, would be for this class to throw an exception!

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -225,26 +225,10 @@ private EsField conflictingTypes(String name, String fullName, FieldCapabilities
                     return unsupported(name, fc);
                 }
-                typesToIndices.computeIfAbsent(type.esType(), _key -> new TreeSet<>()).add(ir.getIndexName());
+                typesToIndices.computeIfAbsent(type.typeName(), _key -> new TreeSet<>()).add(ir.getIndexName());
```
Yeah, I think I just minimized the differences to the original code, but your idea makes sense.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1061,4 +1062,155 @@ public static Expression castStringLiteral(Expression from, DataType target) {
+            imf.getTypesToIndices().forEach((typeName, indexNames) -> {
+                DataType type = DataType.fromTypeName(typeName);
+                TypeResolutionKey key = new TypeResolutionKey(fa.name(), type);
+                if (typeResolutions.containsKey(key)) {
```
Yes, if there are not any convert functions for the types found in the index mappings. Perhaps we have more than one convert function needed, but only one exists in the query. I'm pretty sure I have integration tests (yml tests) that cover this case.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/MultiTypeEsField.java:
```diff
@@ -0,0 +1,116 @@
+
+/**
+ * During IndexResolution it could occur that the same field is mapped to different types in different indices.
+ * The class MultiTypeEfField.UnresolvedField holds that information and allows for later resolution of the field
```
Good point, this comment reflects an earlier version of the class. That is now merged into InvalidMappedField, and not its own class, since we moved QL to ESQL.CORE:
